### PR TITLE
Add InReg attribute to function args

### PR DESCRIPTION
**Pull request by zhanglx13**
Also fixed some bugs in the tuning script

This helps small gemms (~ 10us kernels) a lot. One gemm's execution time drops from 9.5 us to 7.5 us.

### CODE REVIEW

File: lib/Target/LLVMIR/LLVMIRTranslation.cpp [hunk 1]
```diff
@@ -199,6 +199,14 @@ static void amendLLVMFunc(llvm::Function *func, const NVVMMetadata &metadata,
         func->addFnAttr("amdgpu-waves-per-eu", std::to_string(wavesPerEU));
       func->addFnAttr("denormal-fp-math-f32", "preserve-sign");
       func->addFnAttr("amdgpu-unsafe-fp-atomics", "true");
+      for (unsigned I = 0; I < func->arg_size() && I < 16; ++I) {
```
Comment (oplavsic) [inline lib/Target/LLVMIR/LLVMIRTranslation.cpp:202]: What is 16 in  "I < 16" condition?

**Comment by oplavsic** [comment]
@zhanglx13 Do you know exactly what adding this argument achieves? (In terms of how it affects code generation)

**Comment by zhanglx13** [comment]
@oplavsic 
With this attribute, the kernel argument will be preloaded into SGPR instead of memory by the firmware during some kernel launch time. Therefore, at the beginning of the kernel, no `s_load_dword` is required to load kernel args.
In terms of 16, it is the max number of user SGPRs that can be used fore preloading. Maybe I should not hard-code it.

**Comment by zhanglx13** [comment]
@oplavsic So I removed the `I<16` part. 
When there are more than 16 args that want to be preloaded, the firmware will preload the first 16 and the rest still go to memory.

**Review by jayfurmanek — COMMENTED** [review]
This isn't qualified for a specific device, so I assume it's ok (and relevant) to use on any supported AMD device? Have you tried on gfx90a or gfx1100?

**Comment by zhanglx13** [comment]
@jayfurmanek no
Let me try MI250X

**Comment by zhanglx13** [comment]
@jayfurmanek  As expected, no difference on MI250X. 
As long as it does not break anything on MI250, we should be good.

**Comment by jayfurmanek** [comment]
ok great! I'll approve

**Review by jayfurmanek — APPROVED** [review]

**Pull request merged**